Modernize build to support sdists

zbowling opened the pull request:
Replace the two setup setuptools based build process with scikit_build_core.build to call cmake. 

This adds support for sdists releases of this package, git install directly of this repo from pip (submodules may make that hard), and generally simplifies the build process since now all that is required is doing pip install . in the directory or "python -m build" to build the wheel and sdist with no other build steps. This also fixes some cross compiling issues.

This also means that cibuildwheel will work out of the box with github to build this for more architectures and you won't have to maintain as much custom github action CI logic. 

A subset of tests run in CI as part package building now. 

CUDA is optional again which fixes Mac builds. The optional triton kernel is also available as a backend but only when the triton package is also installed in the env.  

Updated documentation and install instructions. 

Add attestation support for uploaded wheels.

Ubospica reviewed (commented):
Hi @zbowling, thanks for the contribution! Modernizing the building process with scikit-build-core and cibuildwheel is definitely helpful for XGrammar. We have planned this for a long time, but there is no bandwidth to do it. So thank you for addressing this important issue!

I updated your PR to make it more suitable for XGrammar's specific situation. I'm also just starting to use these tools, so if there are any problems, feel free to point them out.

We also need to consider how to make CD pipeline work after this migration (as the previous CD depends on setup.py). I will try to work that out later, and contributions are welcome!

Ubospica commented on pyproject.toml:
```diff
@@ -44,3 +128,19 @@ select = ["ALL"]
+# Platform-specific configurations
+[tool.cibuildwheel.linux]
+archs = ["x86_64", "aarch64"]
+# environment = { SKBUILD_BUILD_OPTIONS = "--", CXXFLAGS = "-D_GLIBCXX_USE_CXX11_ABI=0", AUDITWHEEL_PLAT = "manylinux2014_x86_64" }
```
Are these configurations useful?

zbowling replied:
Yeah! Some of this cruft :) I'll clean it up. I sent the draft PR once it started working.

Ubospica commented on pyproject.toml:
```diff
@@ -44,3 +128,19 @@ select = ["ALL"]
+
+[tool.cibuildwheel.macos]
+archs = ["x86_64", "arm64"]
+environment = { SKBUILD_BUILD_OPTIONS = "--", MACOSX_DEPLOYMENT_TARGET = "10.14" }
```
Ditto. Does these configs has any purposes?

zbowling commented:
@Ubospica This is getting closer. There was a big change I made here as part of getting the test to work again with CI is fixing the kernels to work when triton and cuda are not available to backback to CPU. Now I'm able to get all tests to pass on MacOS. Some other non-functional changes to clean up error messages to meet lint warnings got mixed in though when I refactored the API there.

zbowling commented:
The build is green on my repo now! https://github.com/zbowling/xgrammar/actions/runs/13384220521 

We now run a subset of all the unit tests to qualify each wheel during the build (specifically the ones that don't require hugging face tokens). 

This is a bigger PR than I realized.  I fixed several issues with running on MacOS as well. There are a lot of changes overall to fix things to pass tests now on each arch.

zbowling commented:
Oh, one thing here is I upped the minimum supported version to 3.9 instead of 3.8. We can do 3.8 but we have to stop using certain newer typing features in the tests. 

The other cool thing to maybe to investigate is supporting ABI3/py3/PY_LIMITED abi in the wheel and only building for Py39. It doesn't look like we are using anything outside of the stable ABI. This would reduce the shipped number of wheels and speed up CI, making build faster and something that could run quicker on each PR, although it wouldn't have as much of an impact on users since download size is the same.

zbowling commented on python/xgrammar/kernels/__init__.py:
```diff
@@ -1,5 +1,26 @@
+
+apply_token_bitmask_inplace = {
+    "cpu": apply_token_bitmask_inplace_cpu,
+}
```
This moves to a dictionary of the different kernel impls of this function. If cuda isn't available it's not added to then it's not added to the dictionary. These games for triton where if the optional dependency "triton" isn't installed, it's not added to the dictionary as a possible impl.

zbowling commented on python/xgrammar/kernels/apply_token_bitmask_inplace_cpu.py:
```diff
@@ -1,35 +1,60 @@
+BITMASK_CPU_ERROR = "bitmask must be on CPU"
+LOGITS_TYPE_ERROR = "logits must be of type float32"
+BITMASK_TYPE_ERROR = "bitmask must be of type int32"
+MAX_DIMS = 2
```
Some of this was just clean up from my linter to get rid of ruff warnings around const strings in errors and having specific error types.

Ubospica replied:
We can skip the TRY rules for Ruff. There are two reasons:
1. Separating error messages from where they occur could make it harder to maintain
2. We also could throw errors from C++, but the customized exception in Python cannot cover them

zbowling commented on python/xgrammar/matcher.py:
```diff
@@ -121,20 +126,49 @@ def apply_token_bitmask_inplace(
-            apply_token_bitmask_inplace_triton(logits, bitmask, indices)
+        if (
+            os.environ.get("XGRAMMAR_TOKEN_BITMASK_TRITON") == "1"
+            and "triton" in apply_token_bitmask_inplace_impl
```
this logic will also fall back to cuda if triton isn't installed even if you explicitly set XGRAMMAR_TOKEN_BITMASK_TRITON. Alternative we could throw an error.

Ubospica replied:
We can throw an error for this. I have done that in the new commit.

zbowling commented on tests/python/test_grammar_compiler.py:
```diff
@@ -13,6 +13,7 @@
 from xgrammar.testing import _get_allow_empty_rule_ids
 
 
+@pytest.mark.hf_token_required
```
This is a new testing marker I added to flag tests that require a huggingface token with access to the gated models needed in these tests. We skip all these tests at package qualification but I think we still get pretty good coverage. We could run these in CI by passing an HF_TOKEN in an env but then the github runners don't have GPUs so we can't test CUDA paths.

zbowling commented on tests/python/test_grammar_matcher.py:
```diff
@@ -56,15 +82,46 @@ def test_refuse(input_refused: str):
+            4952,
+            127873,
+            128255,
             # fmt: on
```
ruff seems to ignore this comment for some reason.

zbowling commented on tests/python/test_grammar_matcher_ebnf.py:
```diff
@@ -352,7 +520,12 @@ def test_fill_next_token_bitmask(
 
     token_bitmask = xgr.allocate_token_bitmask(1, tokenizer_info.vocab_size)
-    logits_gpu = torch.zeros(tokenizer_info.vocab_size, dtype=torch.float32, device="cuda")
+    device = "cuda" if torch.cuda.is_available() else "cpu"
```
these tests will fallback to a CPU path if CUDA isn't available now which is nice for MacOS

mgorny commented on docs/start/install.rst:
```diff
@@ -83,21 +87,28 @@ This step is useful when you want to make modification or obtain a specific vers
-    # 2. Install pre-commit hooks
+    # 2. Install pre-commit hooks (optional, recommended for contributing to XGrammar)
     pre-commit install
     # 3. build XGrammar core and Python bindings
```
```suggestion
    # 3. build and install XGrammar core and Python bindings
```

mgorny commented on docs/start/install.rst:
```diff
@@ -73,8 +75,10 @@ This step is useful when you want to make modification or obtain a specific vers
         python=3.11
```
```suggestion
        python=3.11 \
        ninja
```

I don't understand why the original instructions combined `conda` and `pip`, but since you're removing the latter part, the dependencies should probably be installed via `conda` instead (in fact, I dare say they should have been like this since the beginning).

Technically, you could also include `pybind11` and `pytorch` here, and use `--no-build-isolation` in pip, then you'd be able to build against GPU-enabled PyTorch from `conda-forge`.

mgorny commented on pyproject.toml:
```diff
@@ -1,19 +1,120 @@
+    "scikit-build-core>=0.10.0",
+    "pybind11>=2.11.0",
+    "wheel",
+    "cmake>=3.18",
+    "ninja",
```
```suggestion
```

scikit-build-core adds these automatically if they are necessary.

mgorny commented on pyproject.toml:
```diff
@@ -1,19 +1,120 @@
+requires = [
+    "scikit-build-core>=0.10.0",
+    "pybind11>=2.11.0",
+    "wheel",
```
```suggestion
```

I don't think `wheel` is used anywhere. Or at least a quick grep doesn't reveal any imports.

maresb commented:
Cross-reference to https://github.com/conda-forge/xgrammar-feedstock/pull/5: Consider removing `pybind11` as a runtime dependency.

Ubospica commented:
Thank a lot @zbowling @maresb! I was too busy the past few days to review it, but I’ll review it tomorrow and get it merged as soon as possible.

Ubospica commented on pyproject.toml:
```diff
@@ -1,19 +1,117 @@
+    "pydantic",
+    "sentencepiece",
+    "tiktoken",
+    "torch>=2.5.0",
```
Support old versions of pytorch
```suggestion
    "torch>=1.10.0",
```

Ubospica commented on pyproject.toml:
```diff
@@ -1,19 +1,117 @@
-strict = true
+[project.optional-dependencies]
+triton = [
+    "triton",
```
Just add triton as a normal dependency
```
"triton; platform_system == 'Linux' and platform_machine == 'x86_64'"
```

zbowling replied:
Sure, but it looks like the code is favoring the direct Cuda version and not the trition version now since you have to set an env to even use triton.

Ubospica commented on pyproject.toml:
```diff
@@ -1,19 +1,117 @@
+# CMake configuration
+cmake.version = "CMakeLists.txt"
+cmake.args = []
+cmake.build-type = "RelWithDebInfo" # Release is Default
```
Setting RelWithDebugInfo as default is fine.
```suggestion
cmake.build-type = "RelWithDebInfo"
```

Ubospica commented on pyproject.toml:
```diff
@@ -44,3 +142,30 @@ select = ["ALL"]
+
+[tool.cibuildwheel]
+build-verbosity = 1
+skip = ["cp36-*", "cp37-*", "cp38-*", "pp*", "*musllinux*", "cp313-manylinux_aarch64", "cp313-macosx_x86_64"] # pypy doesn't play nice with pybind11
```
Why skip cp313-manylinux_aarch64 and cp313-macosx_x86_64?

zbowling replied:
python 3.13 on aarch64 and macosx x86 don't have torch versions for torch 2.5.0+. macos x64 is likely never going to never see any new torch builds for any python version after 3.13. The missing Linux aarch64 build for python 3.13 is mostly because python 3.13 is still just in early support for torch. I'll add a comment for the future.

Ubospica reviewed (commented):
I have finished reviewing the build system section. I will check the linting part tomorrow. Overall, the PR looks great. Thanks!

Just a few questions and points to mention:
1. Can this workflow integrate well with the documentation workflow at [documentation.yaml](https://github.com/mlc-ai/xgrammar/blob/main/.github/workflows/documentation.yaml)?
2. Also, thank you for maintaining the wheel on conda-forge. Is there any part of the conda-forge repository that can be moved into the xgrammar repository? If so, it would be ideal if this could help reduce your maintenance workload.
3. Please use the pre-commit hook to format the code. I noticed some extra blank lines, missing newlines at the end of files, and overly long lines.

zbowling commented on .github/workflows/deploy.yaml:
```diff
@@ -0,0 +1,97 @@
+        - { os: ubuntu-latest,    arch: x86_64,     build: 'cp*-manylinux*' }
+        - { os: ubuntu-24.04-arm, arch: aarch64,    build: 'cp*-manylinux*' }
+        - { os: windows-latest,   arch: AMD64,      build: 'cp*'          }
+        - { os: macos-14,         arch: arm64,      build: 'cp*'          }
```
sure! Yeah since >=2.3 torch dropped support for Mac x64 I wasn't sure how important it but since we ant to support back to torch 1.x that should be easy. There isn't much regression on the older builds but adding it back, the solver should install torch 2.2.2 on Mac x64 and the rest should get torch 2.6 (except for Python 3.13 on aarch64 where there is no official torch binary yet)

zbowling commented:
This should look a bit better. Note I swapped out ruff for black in the pre-commit and fixed up some rules (I was using ruff locally in place of black and it's why it blew up some line wraps but I fixed a few things so now it won't do that which reduces a lot of the changes in this PR.

I also noticed the aarch64 and x86_64 linux builds are lot bigger. Specifically because the cuda kernel that gets compiled now I suspect.

<img width="670" alt="image" src="https://github.com/user-attachments/assets/04dd9848-e0ac-44c7-a206-fd41f0d4318e" />

Ubospica reviewed (commented):
Hi @zbowling, I just finished the review and I think it is close to being complete now!

zbowling commented on pyproject.toml:
```diff
@@ -1,19 +1,107 @@
+
+# Editable install settings
+# Editables are fairly buggy
+# editable.rebuild = true
```
It injects a proxy that tries to rebuild and compile on demand when you import and invoke it. For me, it's not working at all. Something is failing to compile at import time. [Editable native deps are still experimental so I was going to debug that later.](https://scikit-build-core.readthedocs.io/en/latest/configuration/index.html#editable-installs) 

I may still poke at fixing it, but I already bit off a ton more here and figured that could be a follow-up PR once I debugged what is broken there.

Ubospica replied:
It's interesting because the editable installation with
```bash
pip install -e --no-build-isolation .
```
works well for me. The rebuilding can be successfully triggered when the C++ codebase is changed.

But of course we can comment this first since there could still be problems. If you have any further findings, I would appreciate it for bringing it up.

zbowling replied:
Oh nice! Yeah I didn't have time to debug the first pass what was failing.

zbowling replied:
Oh interesting. Editable builds work on my Linux machine but not my Mac. Debugging this stack trace

Ubospica replied:
@zbowling Just wanna ask if this editable installation works now. I feel it is helpful in the real development, so if it has no compatibility issues, we can enable it.

Ubospica commented:
> For me in that conda-forge recipe, I can get around it by changing include paths to include to look at $PREFIX/include/dlpack where dlpack gets install in the dev env instead of 3rdparty/dlpack in the recipe.
> 
> One change we could do is a find_package() module search for dlpack in cmake before falling back to searching in 3rdparty/, and then I wouldn't have to inject any header search paths.

Got it! We can do that in the next PR.

> Picojson was another but this one gets an exception in conda-forge since picojson upstream has been unmaintained for several years and many different ML projects have their own hard forks of it. Googletest isn't used in the conda-forge build so I largely ignore it.

Indeed picojson has been modified a lot by us. I think it can now be treated as part of the codebase rather than an external dependency, so there’s no need to rely on the upstream Conda package.

I think this PR is ready and I will merge it. Thanks a lot for your help @zbowling !

zbowling commented on tests/python/test_grammar_matcher.py:
```diff
@@ -106,13 +96,13 @@ def test_fill_next_token_bitmask(
     """Test accepting token and finding the next token mask."""
-    # fmt: off
     vocab = [
+        # fmt: off
```
I changed these to be external because Ruff ignores them inside the statement. https://docs.astral.sh/ruff/formatter/#format-suppression

Ubospica replied:
Thanks for bringing that up. But since we use black as the formatter instead of ruff, and black supports `fmt: off` by lines, this should be fine as well.
